### ISSUE TITLE
Implement a C^k with k>2 coupling layer with analytic inverse.

### ISSUE THREAD

**Issue by EiffL** [astrodeepnet/sbi_experiments#6]

**Comment by EiffL** [comment]
Ok, so we had a question here: Can we check that for whatever value of a>0 and b \in (0,1), the **end function f** is indeed still a bijection in [0,1]

**Comment by b-remy** [comment]
The demonstration is already in the supplementary information of [Köhler et al.](https://arxiv.org/abs/2110.00351), they demonstrate that sigma[rho](x) is strictly monotonously increasing and bounded, hence bijective. Combination with strictly increasing affine function still gives a bijetor. They also show the smoothness.

Maybe I didn't get the question correctly :-) did you mean experimentally?

**Comment by Justinezgh** [comment]
So they say that : 
<img width="1101" alt="image" src="https://user-images.githubusercontent.com/72011736/145883936-d03c7a2b-74d1-47f7-b967-ce64b2cd7ef9.png">

In our case omega is [-1/2a + b , 1/2a + b] with a>0 and b in [0,1] and I think that considering that ^^^ [-1/2a + b , 1/2a + b] should at least contain [0,1] to have a diffeomorphism on [0,1]. But with these only constraints : a>0 and b in [0,1],  we can choose for instance a = 5 and b = 0.5 ( + c = 0.5) and in this case omega is [0.4 , 0.6] and for rho(x) = exp(-1/x**2) we have this : 

![image](https://user-images.githubusercontent.com/72011736/145889271-001b2c71-b80f-4ee1-9123-f5db487f66bc.png)

**Comment by Justinezgh** [comment]
I think that if we want to use those constraints we have to choose rho in a way that sigma is a bijection on R which is the case for rho(x)=x**3 (I think)

**Comment by EiffL** [comment]
hummmmmm I see, yeah, so it looks like r**k is working as expected,right? is it that the problem is that the rho function should be within [0,1]

**Comment by EiffL** [comment]
hummm yeah, even with r^3 if you put c=10, it's no longer a bijection

**Comment by Justinezgh** [comment]
I think that x**3 is the only one that works. 

rho(x) = x**2
a = 5
b = 0.5
c = 0.001

![image](https://user-images.githubusercontent.com/72011736/145988899-14057844-031e-4227-b612-cc9918d2dd2f.png)


rho(x) = x**5
a = 5
b = 0.5
c = 0.001

![image](https://user-images.githubusercontent.com/72011736/145989159-dd3993d3-0e97-462c-b49b-e059eeea5010.png)

**Comment by Justinezgh** [comment]
So for me we have 2 possibilities : 

- we choose any rho but we have to change the constraints
- we keep the constraints but we have to choose rho in a way that it's a bijection on R

**Comment by EiffL** [comment]
@b-remy  has another explanation that maybe he can explain here. But essentially, despite what is said in the text, f is a bijection from R to I 

So, one way to deal with that would be to add a Scaleshift bijector as input that would rescale I to R before sending it to our bijector

I'm just not sure if in doing so we remove the expressivity of the layer.... To be tested. So Justine can you try adding such a bijector in your coupling layer? So that the input and output of the layer remain in [0,1]

**Comment by Justinezgh** [comment]
The constraints if we want to have [0,1] in omega 

![6DC602B8-46FF-4ED5-8170-60BB2DBFA002](https://user-images.githubusercontent.com/72011736/146006880-9b97469a-5cd1-4923-88e0-8910400c6de8.jpeg)

![image](https://user-images.githubusercontent.com/72011736/146007593-a86c733a-f983-4fc2-997a-d0c32027ce4d.png)

**Comment by Justinezgh** [comment]
I'm not sure about the bijection f from R to [0,1]. To have a bijection on R, sigma has to be a bijection on R too even if [0,1] is in omega : 

rho(x) = x**4
![image](https://user-images.githubusercontent.com/72011736/146025670-53ce5e94-a57c-4bf0-aa1e-7272c30dd542.png)


and f(x)  with x in R can't be in [0,1] because of the c*x

**Comment by Justinezgh** [comment]
> The constraints if we want to have [0,1] in omega
> 
> ![6DC602B8-46FF-4ED5-8170-60BB2DBFA002](https://user-images.githubusercontent.com/72011736/146006880-9b97469a-5cd1-4923-88e0-8910400c6de8.jpeg)
> 
> ![image](https://user-images.githubusercontent.com/72011736/146007593-a86c733a-f983-4fc2-997a-d0c32027ce4d.png)

So to keep it simple we can say that b is in [0.1 , 0.9] and a is in ]0 , 0.5] (because 1/(2(1-b)) with b = 0.1 is 0.56) idk :/
The thing is that if we use a conditional activation function I think that we will have the same pb as with relu

**Comment by EiffL** [comment]
hummmmmm its really not clear to me, what you suggest looks like it should work, but I'm surprised it's not mentioned at all in the paper. Can we try to find their source code somewhere and check their implementation?

**Comment by Justinezgh** [comment]
Benjamin you mean for g ? I'm not sure I understand

**Comment by Justinezgh** [comment]
I can't find their code :/

**Comment by b-remy** [comment]
> Benjamin you mean for g ? I'm not sure I understand

I meant f, it is the same for g since g is a bijective function only from [-1/2/a+b, 1/2/a+b] to [0,1], actually I did add anything useful to the discussion ^^ but now I get your point.

To understand what's going on outside [0,1], you can easily show that if you want sigma to be bijective for all k (odd or even), you need to take x in [0,1]. All R works for odd k, but only x in [0,1] for even k (by considering the derivative of sigma > 0).

**Comment by Justinezgh** [comment]
Axel helped me to find this : https://github.com/noegroup/bgflow/pull/28

**Comment by aboucaud** [comment]
So what should we look at in this repo ?
The coupling flow is there https://github.com/noegroup/bgflow/blob/main/bgflow/nn/flow/coupling.py#L133

**Comment by EiffL** [comment]
https://github.com/noegroup/bgflow/blob/191465c29325d07d371c3bcc2e380716acf4666d/bgflow/nn/flow/transformer/compact/compact.py#L195

**Comment by EiffL** [comment]
In fact.... this looks like the piece of code we are interested in:
https://github.com/noegroup/bgflow/blob/52f9e32a8d3b432ce559fa857ee7dc17997e9b96/bgflow/nn/flow/transformer/sigmoid.py#L62
I would say that here this is the definition of this sigmoid coupling layer, the code I have in the post above that where they define the ramp functions.

After some reverse engineering, I think I got it!
Here is the translation table: a = sigma, b = mu, c = min_density
And here is the coupling function:
```python
def affine_sigmoid_transform(x, a, b, c, ramp=rho):

  b = (b - 1./(2*a))

  # Here we apply a rescaling by a and b first, before rho
  diff = x - b
  zs = jnp.stack([diff, -b*jnp.ones_like(x), 1 - b*jnp.ones_like(x)],axis=0)
  zs = zs * a

  def sigma(x):
    return rho(x) / (rho(x) + rho(1 - x))

  y, y0, y1 = sigma(zs)

  # This is the equivalent of f
  y = (y - y0)/ (y1 - y0)

  y = y*(1-c) + c *x
  
  return y
```
![image](https://user-images.githubusercontent.com/861591/146102812-5dc96c6c-3dca-48b9-8990-1712bb11f86c.png)
Here is the result for an exponential function and the function defined above \o/

**Comment by EiffL** [comment]
hum actually not completely solved, it doesnt work for a > 1

**Comment by EiffL** [comment]
need to check carefully how this function restricts the values of a,b,c that come from the NN: https://github.com/noegroup/bgflow/blob/52f9e32a8d3b432ce559fa857ee7dc17997e9b96/bgflow/nn/flow/transformer/sigmoid.py#L78

**Comment by Justinezgh** [comment]
> need to check carefully how this function restricts the values of a,b,c that come from the NN: https://github.com/noegroup/bgflow/blob/52f9e32a8d3b432ce559fa857ee7dc17997e9b96/bgflow/nn/flow/transformer/sigmoid.py#L78

ok I think that I'll need to think about it a bit more ^^' but I did that : https://colab.research.google.com/drive/1mRGwrsJl0ldW9wul3I1DsdfYjC1IjcOO?usp=sharing

**Comment by EiffL** [comment]
Lol, yep, I think that's the black magic we had been looking for :-D Could you try to modify the code in https://github.com/astrodeepnet/sbi_experiments/pull/10 accordingly? Maybe only the notebook needs to be changed... I'm not sure.

**Comment by EiffL** [comment]
Ok this is great @Justinezgh if you want to add your analytic inverse code, which would be nice for comparison, you can open a Pull Request for this, I'll be happy to review it ;-)

**Comment by Justinezgh** [comment]
> In fact.... this looks like the piece of code we are interested in: https://github.com/noegroup/bgflow/blob/52f9e32a8d3b432ce559fa857ee7dc17997e9b96/bgflow/nn/flow/transformer/sigmoid.py#L62 I would say that here this is the definition of this sigmoid coupling layer, the code I have in the post above that where they define the ramp functions.
> 
> After some reverse engineering, I think I got it! Here is the translation table: a = sigma, b = mu, c = min_density And here is the coupling function:
> 
> ```python
> def affine_sigmoid_transform(x, a, b, c, ramp=rho):
> 
>   b = (b - 1./(2*a))
> 
>   # Here we apply a rescaling by a and b first, before rho
>   diff = x - b
>   zs = jnp.stack([diff, -b*jnp.ones_like(x), 1 - b*jnp.ones_like(x)],axis=0)
>   zs = zs * a
> 
>   def sigma(x):
>     return rho(x) / (rho(x) + rho(1 - x))
> 
>   y, y0, y1 = sigma(zs)
> 
>   # This is the equivalent of f
>   y = (y - y0)/ (y1 - y0)
> 
>   y = y*(1-c) + c *x
>   
>   return y
> ```
> 
> ![image](https://user-images.githubusercontent.com/861591/146102812-5dc96c6c-3dca-48b9-8990-1712bb11f86c.png) Here is the result for an exponential function and the function defined above \o/



But this function is exactly the same that the one describe in the paper right ? And I'm not sure that those constraints :
```python 
    log_a_bound=4
    min_density_lower_bound=1e-4

    log_a   = jnp.exp(jax.nn.tanh(hk.Linear(output_units)(net)) * log_a_bound)
    b   = jax.nn.sigmoid(hk.Linear(output_units)(net))
    c   = min_density_lower_bound + jax.nn.sigmoid(hk.Linear(output_units)(net)) * (1 - min_density_lower_bound)
```

solved the pb.
Just to be sure, I tried the nf ([https://github.com/astrodeepnet/sbi_experiments/blob/ramp_bijector/notebooks/score_matching/NF_implicit.ipynb](url)) with rho(x) = x^6 and rho(x) = exp(-1/(x^2) (f is not a bijection on [0,1] for these rho :

![image](https://user-images.githubusercontent.com/72011736/146694189-03253160-8fa7-4cc6-badd-1ef0ee965c4e.png)
![image](https://user-images.githubusercontent.com/72011736/146694211-c1016416-2faf-4bfe-aff2-21b4e55a290c.png)

) and the nf crashed. I think that it works for rho(x) = x^5 because as Benjamin mentioned f is a bijection if k is odd. (Or maybe I'm just super confused ^^')

**Comment by EiffL** [comment]
Do I don't know if you have tried it as well, but here is a notebook to try the different bijectors from bgflow for comparison:
https://colab.research.google.com/drive/1ar7QkUSPxMfMnM4P8SC4RpOp2MK2fgO3?usp=sharing

**Comment by Justinezgh** [comment]
This is the reason why I have always 'nan' in my loss : 
![image](https://user-images.githubusercontent.com/72011736/147231624-408dccff-e304-4ca7-ac3e-e883834673f9.png)
 And I don't think that I can have a better inverse with sympy

**Comment by Justinezgh** [comment]
> So I don't know if you have tried it as well, but here is a notebook to try the different bijectors from bgflow for comparison: https://colab.research.google.com/drive/1ar7QkUSPxMfMnM4P8SC4RpOp2MK2fgO3?usp=sharing

I will look at it, thanks :)

**Comment by EiffL** [comment]
Hummmm I see.... Well I think maybe there is something wrong with our bijector not being exactly what it should be, which could explain why your inverse fails outside a given range.

But more generally, now that we have an implicit inverse that **seems** to work maybe we don't need to spend too much time on the analytic inverse, which we know is going to be limited to some special cases.

Whatever the way to achieve it, our goal is to have a normalizing flow that we can train from its score.

**Comment by Justinezgh** [comment]
yup, I tried with strong constraints on a and the results are not good at all since we lose a lot of expressivity. Even if I increase the number of coupling layers

**Comment by EiffL** [comment]
I think.... we can close this issue now, right?
